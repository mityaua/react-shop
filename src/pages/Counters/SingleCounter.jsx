import { createUseStyles } from 'react-jss';

const useStyles = createUseStyles({
  counter: {
    padding: 10,
  },
  buttons: {
    display: 'flex',
  },
});

const SingleCounter = ({
  step,
  value,
  onSetStep,
  onDecrement,
  onIncrement,
}) => {
  const classes = useStyles();

  const handleChangeStep = event => onSetStep(Number(event.target.value));
  const handleDecrement = () => onDecrement(step);
  const handleIncrement = () => onIncrement(step);

  return (
    <div className={classes.counter}>
      <h4>Single Counter Redux</h4>

      <label>
        <span>Step</span>
        <select value={step} onChange={handleChangeStep}>
          <option value="1">1</option>
          <option value="5">5</option>
          <option value="10">10</option>
          <option value="25">25</option>
          <option value="50">50</option>
        </select>
      </label>

      <div className={classes.buttons}>
        <button onClick={handleDecrement}>-</button>
        <p data-testid="value">{value}</p>
        <button data-testid="increment-button" onClick={handleIncrement}>
          +
        </button>
      </div>
      <hr />
    </div>
  );
};

export default SingleCounter;

export const sum = (a, b) => a + b;
