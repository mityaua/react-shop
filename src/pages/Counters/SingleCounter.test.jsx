// Файл с примером тестирования
// для запуска теста используем // it() или test()
import { render } from '@testing-library/react';
import SingleCounter, { sum } from './SingleCounter';

// Тестирование функций
describe('sum', () => {
  it('sum 1 + 1 = 2', () => {
    expect(sum(1, 1)).toBe(2);
  });

  it('wrong test example', () => {
    expect(sum(5, 5)).toBe(10);
  });

  it('NaN check', () => {
    expect(sum(111)).toBeNaN();
  });

  it('empty params check', () => {
    expect(sum()).toBeNaN();
  });
});

// Тест компонента SingleCounter
describe('SingleCounter', () => {
  // Тест рендера контейнера
  it('snapshot test', () => {
    const { container } = render(<SingleCounter value={555} />);

    expect(container).toMatchSnapshot();
  });

  // Доступ к элементу DOM
  it('value test', () => {
    const { getByTestId } = render(<SingleCounter value={555} />);

    expect(getByTestId('value').textContent).toBe('555');
  });

  // Тест вызова функции на кнопке
  it('test increment button func', () => {
    const onIncrement = jest.fn();

    const { getByTestId } = render(<SingleCounter onIncrement={onIncrement} />);

    getByTestId('increment-button').click();
    getByTestId('increment-button').click();
    getByTestId('increment-button').click();

    expect(onIncrement).toBeCalled();
    expect(onIncrement).toBeCalledTimes(3);
  });
});

// Для более сложных функций можно использовать Mock-функции https://jestjs.io/ru/docs/mock-function-api
