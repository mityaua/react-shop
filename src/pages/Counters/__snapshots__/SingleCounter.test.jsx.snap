// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleCounter snapshot test 1`] = `
<div>
  <div
    class="counter-0-2-1"
  >
    <h4>
      Single Counter Redux
    </h4>
    <label>
      <span>
        Step
      </span>
      <select>
        <option
          value="1"
        >
          1
        </option>
        <option
          value="5"
        >
          5
        </option>
        <option
          value="10"
        >
          10
        </option>
        <option
          value="25"
        >
          25
        </option>
        <option
          value="50"
        >
          50
        </option>
      </select>
    </label>
    <div
      class="buttons-0-2-2"
    >
      <button>
        -
      </button>
      <p
        data-testid="value"
      >
        555
      </p>
      <button
        data-testid="increment-button"
      >
        +
      </button>
    </div>
    <hr />
  </div>
</div>
`;
